core/src/main/java/org/elasticsearch/search/slice/DocValuesSliceQuery.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.slice;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.SortedNumericDocValues;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.search.RandomAccessWeight;
+import org.apache.lucene.util.Bits;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.index.fielddata.AtomicFieldData;
+import org.elasticsearch.index.fielddata.AtomicNumericFieldData;
+import org.elasticsearch.index.fielddata.IndexFieldData;
+import org.elasticsearch.index.fielddata.SortedBinaryDocValues;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * Query that uses the doc_values of a field to slice the documents
+ */
+final class DocValuesSliceQuery extends SliceQuery {
+    private final IndexFieldData<?> indexFieldData;
+
+    public DocValuesSliceQuery(Slice slice, IndexFieldData indexFieldData) {
+        super(slice);
+        this.indexFieldData = indexFieldData;
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        if (other instanceof DocValuesSliceQuery) {
+            return super.equals(other) &&
+                indexFieldData == ((DocValuesSliceQuery) other).indexFieldData;
+        }
+        return false;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(getSlice(), indexFieldData.hashCode());
```

Comment:
s/getSlice()/super.hashCode()/ to be sure to fold all the hashcode components of the parent class too?